gotran/codegeneration/solvercomponents.py

```diff
@@ -274,6 +274,29 @@ def __init__(self, ode, function_name="forward_explicit_euler", params=None):
         results, body_expressions = self._body_from_results(**results)
         self.body_expressions = self._recreate_body(body_expressions, **results)
 
+def fraction_numerator_is_nonzero(expr):
+    """Perform a very cheap check to detect if a fraction is definitely non-zero."""
+    expr_type = type(expr)
+    if expr_type == sp.Pow:
+        # check if the expression is on the form a**-1
+        a, b = expr.args
+        if type(b) is sp.numbers.NegativeOne:
+            return True
+        else:
+            # we won't do any further checks
+            return False
+    elif expr_type == sp.Mul:
+        args = expr.args
+        potentially_nonzero_args = []
+        for e in args:
+            if not (len(e.free_symbols) == 0 and e.is_nonzero):
+                potentially_nonzero_args.append(e)
+
+        if len(potentially_nonzero_args) > 1:
+            return False
+        return fraction_numerator_is_nonzero(*potentially_nonzero_args)
```

Comment:
I don't understand this completely. You find the arguments that are nonzero. If you find more than one return False otherwise make a recursive call on the arguments of the unpacked list. What happens if `potentially_nonzero_args` is empty at this points? `fraction_numerator_is_nonzero` takes only one arguments so perhaps we should check that the list is not empty?

Reply:
I also found it confusing when reading my own code two weeks later. I've hopefully made it a bit more clear now.